File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        int len = nums.size();
        unordered_map<int, int> windowElements;
        long long res = 0LL;
        long long currSum = 0LL;
        for (int beg = 0, end = 0; end < len; end++) {
            currSum += nums[end];
            windowElements[nums[end]]++;

            int currLen = end - beg + 1;
            while (currLen > k) {
                currSum -= nums[beg];
                windowElements[nums[beg]]--;

                if (windowElements[nums[beg]] == 0) {
                    windowElements.erase(nums[beg]);
                }
                beg++;
                currLen = end - beg + 1;
            }

            if (windowElements.size() == k) {
                res = max(res, currSum);
            }
        }

        return res;
    }
};

File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        int len = nums.length;

        long currSum = 0;
        int left = 0;
        int right = 0;
        Map<Integer, Integer> freq = new HashMap<>();
        while (right < k - 1) {
            int curr = nums[right];
            currSum += curr;
            int count = freq.getOrDefault(curr, 0);
            freq.put(curr, count + 1);
            right += 1;
        }

        long ans = 0;
        for ( ; right < len; right++, left++) {
            int curr = nums[right];
            currSum += curr;
            int count = freq.getOrDefault(curr, 0);
            freq.put(curr, count + 1);

            if (freq.size() == k) {
                ans = Math.max(ans, currSum);
            }

            int windowBeg = nums[left];
            currSum -= windowBeg;
            count = freq.get(windowBeg);
            freq.put(windowBeg, count - 1);
            if ((count - 1) == 0) {
                freq.remove(windowBeg);
            }
        }   

        return ans;
    }
}